refactor(increments-service): update routes and handlers for page views functionality

Yedidyar opened the pull request:
- Change route prefix from "/increments" to "/page-views" in the service and tests
- Implement new handler for batch increment requests with "/multi" endpoint
- Update existing increment logic to support multiple page views in a single request
- Adjust integration tests to reflect new endpoint structure and functionality

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->

## Summary by CodeRabbit

* **New Features**
  * Added a new endpoint to support batch creation and updating of multiple page views in a single request.
* **Bug Fixes**
  * Improved error handling for invalid timestamp formats in batch operations.
* **Refactor**
  * All date and time handling for page views now use UTC for consistency.
  * The route prefix for page view endpoints has changed from "/increments" to "/page-views".
* **Tests**
  * Added integration tests for the new batch page views endpoint, including edge cases and aggregation logic.

<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Commit:
refactor(increments-service): update routes and handlers for page views functionality

- Change route prefix from "/increments" to "/page-views" in the service and tests
- Implement new handler for batch increment requests with "/multi" endpoint
- Update existing increment logic to support multiple page views in a single request
- Adjust integration tests to reflect new endpoint structure and functionality

Yedidyar commented:
@coderabbitai review

Yedidyar commented:
@coderabbitai give me places that I should improve the code

tomer-shvadron commented on analytics-project/increments-service/handlers/index.ts:
```diff
@@ -13,6 +17,7 @@ export async function createOrUpdateSingleIncrementHandler(
     const { page, timestamp } = request.body;
 
     await request.server.incrementsService.incrementPage(page, timestamp);
+    return reply.status(200).send({ success: true });
```
```suggestion
    return reply.status(200).send();
```

tomer-shvadron commented on analytics-project/increments-service/handlers/index.ts:
```diff
@@ -26,6 +31,28 @@ export async function createOrUpdateSingleIncrementHandler(
+    const pageData = request.body;
+
+    await request.server.incrementsService.incrementMultiplePages(pageData);
+    return reply.status(200).send({ success: true });
```
```suggestion
    return reply.status(200).send();
```

tomer-shvadron commented on analytics-project/increments-service/handlers/index.ts:
```diff
@@ -26,6 +31,28 @@ export async function createOrUpdateSingleIncrementHandler(
+  try {
+    const pageData = request.body;
+
+    await request.server.incrementsService.incrementMultiplePages(pageData);
```
```suggestion
    await request.server.incrementsService.incrementMultiplePages(request.body);
```

tomer-shvadron commented on analytics-project/increments-service/integration.test.ts:
```diff
@@ -87,4 +87,149 @@ describe("Increments Service API Integration Tests", () => {
+      expect(altmanResult.rows[1].views_count).toBe("200");
+      expect(altmanResult.rows[1].hour).toBe(1);
+      expect(altmanResult.rows[2].views_count).toBe("405");
+      expect(altmanResult.rows[2].hour).toBe(2);
```
Let's compare objects, checking it like this is tedious and ugly

tomer-shvadron commented on analytics-project/increments-service/integration.test.ts:
```diff
@@ -87,4 +87,149 @@ describe("Increments Service API Integration Tests", () => {
+      expect(muskResult.rows[0].hour).toBe(0);
+      expect(muskResult.rows[1].views_count).toBe("654");
+      expect(muskResult.rows[1].hour).toBe(1);
+    });
```
Same comment as above

tomer-shvadron commented on analytics-project/increments-service/integration.test.ts:
```diff
@@ -87,4 +87,149 @@ describe("Increments Service API Integration Tests", () => {
+      expect(result.rows[0].hour).toBe(0);
+      expect(result.rows[0].views_count).toBe("600");
+
+      // Hour 1 should have the single value: 150
```
I hate these comments that the AI adds to our code. Makes it looks less professional as they are unnecessary and add no value

tomer-shvadron commented on analytics-project/increments-service/repositories/increments.ts:
```diff
@@ -20,17 +20,33 @@ export class IncrementsRepository {
   }
 
+  async #executeIncrementQuery(
+    session: any,
```
Great catch but bad suggestion - `pg` exports necessary types

tomer-shvadron commented on analytics-project/increments-service/repositories/increments.ts:
```diff
@@ -46,4 +62,137 @@ export class IncrementsRepository {
+        paramIndex += 5;
+      }
+
+      const sql = `
+        INSERT INTO page_views (id, name, date, hour, views_count) 
+        VALUES ${placeholders.join(", ")}
+        ON CONFLICT (name, date, hour) 
+        DO UPDATE SET views_count = page_views.views_count + EXCLUDED.views_count
+      `;
+
+      await session.query(sql, values);
+
+      const successfulResults = pageIncrements.map((increment) => ({
+        page: increment.page,
+        date: increment.date,
+        hour: increment.hour,
+        success: true,
+      }));
+
+      return {
+        successful: successfulResults,
+        failed: [],
+        totalProcessed: pageIncrements.length,
+        successCount: pageIncrements.length,
+        errorCount: 0,
+      };
```
What do we think about atomicity in this section? What should we do if X our of Y rows fail?

Yedidyar replied:
In this case there is no partial failures

tomer-shvadron replied:
Let's ask Ron about it

tomer-shvadron commented on analytics-project/increments-service/services/increments.service.ts:
```diff
@@ -5,9 +5,71 @@ export class IncrementsService {
     const value = 1;
 
-    await this.incrementsRepository.incrementPage(page, date, hour, value);
+    const dateOnly = date.toISOString().split("T")[0];
```
This should either be wrapped with a try-catch or we need to add validation over the date string that we get in the report

Yedidyar replied:
I don't understand this comment?
Why it should be wrapped with try catch?

tomer-shvadron replied:
Thinking about it, we have no validation of the timestamp. That means that when we create the Date object from the timestamp string, we get an `Invalid Date` object, and then `getUTCHours` will cause an error.
So - validation of the timestamp (Regex, I assume) and wrapping with a try-catch to be on the safe side? WDYT?

Yedidyar commented:
> Thinking about it, we have no validation of the timestamp. That means that when we create the Date object from the timestamp string, we get an `Invalid Date` object, and then `getUTCHours` will cause an error.
> So - validation of the timestamp (Regex, I assume) and wrapping with a try-catch to be on the safe side? WDYT?

If We should add validation for all of the endpoints
@tomer-shvadron

Commit:
refactor(increments-service): enhance error handling and validation in increment functionality

- Introduce constants for HTTP status codes and error messages for better maintainability
- Implement detailed validation for page identifiers and timestamp formats in service methods
- Update integration tests to cover new validation scenarios, including empty requests and negative increment values
- Refactor response handling in handlers to return specific error messages based on validation failures

Commit:
refactor(increments-service): simplify validation logic and enhance error handling

- Remove redundant validation functions for single and multiple increments
- Integrate validation directly into the request handling logic for improved clarity
- Update integration tests to reflect changes in error responses for empty requests and invalid timestamps
